core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/container/ContainerManager.scala

```diff
@@ -395,64 +466,80 @@ object ContainerManager {
+            val updatedInvokers =
+              updateInvokerMemory(scheduledPair.invokerId, msg.whiskActionMetaData.limits.memory.megabytes, invokers)
+            (scheduledPair :: pairs, updatedInvokers)
           } else {
-            sendState(
-              FailedCreationJob(
-                msg.creationId,
-                msg.invocationNamespace,
-                msg.action,
-                msg.revision,
-                NoAvailableInvokersError,
-                s"No available invokers."))
-            (pairs, candidates)
+            val scheduledPair =
+              ScheduledPair(msg, invokerId = None, Some(NoAvailableInvokersError), Some(s"No available invokers."))
+            (scheduledPair :: pairs, invokers)
           }
         } else {
           val wantedInvokers = candidates.filter(health => requiredResources.toSet.subsetOf(health.id.tags.toSet))
           if (wantedInvokers.nonEmpty) {
-            chooseInvokerFromCandidates(wantedInvokers, invokers, pairs, msg)
+            val scheduledPair = chooseInvokerFromCandidates(wantedInvokers, msg)
+            val updatedInvokers =
+              updateInvokerMemory(scheduledPair.invokerId, msg.whiskActionMetaData.limits.memory.megabytes, invokers)
+            (scheduledPair :: pairs, updatedInvokers)
           } else if (resourcesStrictPolicy) {
-            sendState(
-              FailedCreationJob(
-                msg.creationId,
-                msg.invocationNamespace,
-                msg.action,
-                msg.revision,
-                NoAvailableResourceInvokersError,
-                s"No available invokers with resources $requiredResources."))
-            (pairs, candidates)
+            val scheduledPair =
+              ScheduledPair(
+                msg,
+                invokerId = None,
+                Some(NoAvailableResourceInvokersError),
+                Some(s"No available invokers with resources $requiredResources."))
+            (scheduledPair :: pairs, invokers)
           } else {
             val (noTaggedInvokers, taggedInvokers) = candidates.partition(_.id.tags.isEmpty)
             if (noTaggedInvokers.nonEmpty) { // choose no tagged invokers first
-              chooseInvokerFromCandidates(noTaggedInvokers, invokers, pairs, msg)
+              val scheduledPair = chooseInvokerFromCandidates(noTaggedInvokers, msg)
+              val updatedInvokers =
+                updateInvokerMemory(scheduledPair.invokerId, msg.whiskActionMetaData.limits.memory.megabytes, invokers)
+              (scheduledPair :: pairs, updatedInvokers)
             } else {
               val leftInvokers =
                 taggedInvokers.filterNot(health => requiredResources.toSet.subsetOf(health.id.tags.toSet))
-              if (leftInvokers.nonEmpty)
-                chooseInvokerFromCandidates(leftInvokers, invokers, pairs, msg)
-              else {
-                sendState(
```

Comment:
This `schedule` method also had side effects like this one.
It was sending a state. Since it makes it harder to test this method, I also removed this kind of side effect too.
Now the `schedule` method only schedules invokers for the given cold creations.